lib/synapse/service_watcher/base.rb

```diff
@@ -102,15 +120,16 @@ def validate_discovery_opts
       log.warn "synapse: warning: a stub watcher with no default servers is pretty useless" if @default_servers.empty?
     end
 
-    def filter_backends_by_label(backends, label_filter)
-      filtered_backends = []
-      backends.each do |backend|
+    def backends_filtered_by_labels
+      filtered_backends = @backends.select do |backend|
         backend_labels = backend['labels'] || {}
-        if label_filter['condition'] == 'equals' and backend_labels[label_filter['label']] == label_filter['value']
-          filtered_backends << backend
+        @label_filters.all? do |label_filter|
+          (label_filter['condition'] == 'equals' &&
+            backend_labels[label_filter['label']] == label_filter['value']) ||
+          (label_filter['condition'] == 'not-equals' &&
```

Comment:
Food for thought: So right now, if a node is missing a label, it makes sense for the `equals` label condition to fail (we're missing information, so it can't be `equal` unless we are expecting a `nil` or missing value). So there's no confusion here for the `equals` filter.

For `not-equals`, this scenario may cause confusion. When a backend node is missing the label value altogether, it will always pass the `not-equals` label condition. Right now, there's no chance of the labels information in Nerve being flaky as they're statically configured. This could be a concern in the future when label information is dynamic and labels could drop in/out of existence.